### PR TITLE
Prestwich/dyn sol error

### PR DESCRIPTION
**Pull request by prestwich**
## Motivation

Work on bringing dyn-abi up to feature parity with sol-types

## Solution

Add `DynSolError` to match `DynSolType` and `DynSolEvent`

## PR Checklist

- [ ] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/dyn-abi/src/dynamic/error.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
+use crate::{DynSolType, DynSolValue, Error, Result};
+use alloc::vec::Vec;
+use alloy_primitives::{keccak256, Selector};
+
+const PANIC_SELECTOR: Selector = Selector::new([0x4e, 0x48, 0x7b, 0x71]);
+const REVERT_SELECTOR: Selector = Selector::new([0x08, 0xc3, 0x79, 0xa0]);
```
Comment (Evalir) [inline crates/dyn-abi/src/dynamic/error.rs:6]: Can we add some doc comments with the string version of these? (`Panic(uint256)` & `Error(string)`)

Comment (DaniPopes) [reply crates/dyn-abi/src/dynamic/error.rs:6]: Could also be `alloy_sol_types::Revert::SELECTOR`

File: crates/dyn-abi/src/dynamic/error.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+/// A dynamic ABI error.
+///
+/// This is a representation of a Solidity error, which can be used to decode
+/// error events.
+#[derive(Debug, Clone, PartialEq)]
+pub struct DynSolError {
+    /// Error selector.
+    pub(crate) selector: Selector,
+    /// Error body types.
+    pub(crate) body: DynSolType,
```
Comment (DaniPopes) [inline crates/dyn-abi/src/dynamic/error.rs:17]: `Vec<DynSolType>`?

File: crates/dyn-abi/src/dynamic/error.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    pub const fn new(selector: Selector, body: DynSolType) -> Self {
+        Self { selector, body }
+    }
+
+    /// Error selector is the first 4 bytes of the keccak256 hash of the error
+    /// declaration.
+    pub const fn selector(&self) -> Selector {
+        self.selector
+    }
+
```
Comment (DaniPopes) [inline crates/dyn-abi/src/dynamic/error.rs:56]: `body` getter

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/dyn-abi/src/dynamic/error.rs (hunk 2 above)
Comment (prestwich) [reply crates/dyn-abi/src/dynamic/error.rs:20]: `DynSolEvent` uses `DynSolType` for body. could change both?

**Pull request merged**